### ISSUE TITLE
Missing GPU case/optimization for some axis patterns in GpuCAReduce

### ISSUE THREAD

**Issue by lamblin** [Theano/Theano#2254]
Reductions over axes `(0, 1, 3)` for a 4D tensor (at least) are not transferred to GPU.

```python
>>> import theano
>>> import numpy as np

>>> x = theano.shared(np.ones((2,3,4,5), dtype=theano.config.floatX))
>>> f = theano.function([], theano.tensor.sum(x, (0, 1, 3)))
>>> theano.printing.debugprint(f, print_type=True)
Sum{axis=[0, 1, 3], acc_dtype=float64} [@A] <TensorType(float32, vector)> ''   1
 |HostFromGpu [@B] <TensorType(float32, 4D)> ''   0
   |<CudaNdarrayType(float32, 4D)> [@C] <CudaNdarrayType(float32, 4D)>
```
It works for other axes, like `(0, 2, 3)`:
```python
>>> g = theano.function([], theano.tensor.sum(x, (0, 2, 3)))
>>> theano.printing.debugprint(g, print_type=True)
HostFromGpu [@A] <TensorType(float32, vector)> ''   1
 |GpuCAReduce{add}{0,1,1,1} [@B] <CudaNdarrayType(float32, vector)> ''   0
   |<CudaNdarrayType(float32, 4D)> [@C] <CudaNdarrayType(float32, 4D)>
```

or, with a more complex expression, for `(0, 1, 2)`: 
```python
>>> h = theano.function([], theano.tensor.sum(x, (0, 1, 2)))
>>> theano.printing.debugprint(h, print_type=True)
HostFromGpu [@A] <TensorType(float32, vector)> ''   8
 |GpuCAReduce{add}{1,0} [@B] <CudaNdarrayType(float32, vector)> ''   7
   |GpuReshape{2} [@C] <CudaNdarrayType(float32, matrix)> ''   6
     |<CudaNdarrayType(float32, 4D)> [@D] <CudaNdarrayType(float32, 4D)>
     |MakeVector [@E] <TensorType(int64, vector)> ''   5
       |Elemwise{Mul}[(0, 0)] [@F] <TensorType(int64, scalar)> ''   4
       | |Shape_i{2} [@G] <TensorType(int64, scalar)> ''   3
       | | |<CudaNdarrayType(float32, 4D)> [@D] <CudaNdarrayType(float32, 4D)>
       | |Shape_i{0} [@H] <TensorType(int64, scalar)> ''   2
       | | |<CudaNdarrayType(float32, 4D)> [@D] <CudaNdarrayType(float32, 4D)>
       | |Shape_i{1} [@I] <TensorType(int64, scalar)> ''   1
       |   |<CudaNdarrayType(float32, 4D)> [@D] <CudaNdarrayType(float32, 4D)>
       |Shape_i{3} [@J] <TensorType(int64, scalar)> ''   0
         |<CudaNdarrayType(float32, 4D)> [@D] <CudaNdarrayType(float32, 4D)>
```

**Comment by nouiz** [comment]
In the new back-end we have a generic case that support all combination. This would solve this.

**Comment by bouthilx** [comment]
@nouiz Should we implement 101 pattern or use the new back-end for a generic solution? I'm not quite sure what you mean by the new back-end.

**Comment by abergeron** [comment]
I think the point is to make the dimshuffles/reshapes to make it work in the old backend.  The new one already has a generic solution which covers all patterns.

**Comment by nouiz** [comment]
I think we should use that in the new back-end too. But we should first
confirm if it work or not in the new back-end for that case and if it work,
is there a speed difference or not. If the new back-end is slower, we
should implement that in the new back-end.

**Comment by variousvirgil** [comment]
Could you please tell me how to add pattern "10111" manually?

**Comment by nouiz** [comment]
This patter should be handled automatically, as when the base pattern don't
exist, we try other version by collapsing dimensions. In your case, this
end up being 101. Which is the only 3nd pattern we currently don't support.
Someone is working on this. So when done your case should get support
automatically.

But for now, you can replace:

a_5d_tensor.sum(axis=[0,2,3,4])

by:
a_5d_tensor.flatten(4).sum(axis=[0,2,3])

**Comment by variousvirgil** [comment]
@nouiz Thanks for your quick reply!
My problem is that the ```sum``` op is auto generated for computing gradient so that I don't know how to replace the operation.

**Comment by nouiz** [comment]
So you need to wait for someone to finish this. Some is working on this.
I'm not sure when it will be done, but expect at worst a few weeks.

If you want to make it yourself to get it faster, you need to modify the
file sandbox/cuda/opt.py there is a function that move the op CAReduce to
the GPU. You can make it detect your pattern and replace it with the CPU or
GPU graph that do what you need.

**Comment by variousvirgil** [comment]
@nouiz Thanks!

**Comment by abergeron** [comment]
Closed by #3696
